G2 visual followup

jasmussen opened the pull request:
This PR fixes #20575 and addresses feedback, plus some more.

- It fixes a "busy" gradient when saving reusable blocks.
- It polishes the active state for alignments and block navigation.
- It removes the triangle indicator from the block toolbar, because for the time being at least, we can't offset the toolbar 48px to the left in a consistent way, so its initial purpose is not yet met.
- It removes the gray flashing below new paragraphs, as it hasn't proven itself valuable.
- It adds animation to the toggle buttons in the block toolbar.
- It defaults all popovers to "down and right", bringing consistency.
- It adds 12px spacing between popover and opener, making sure popovers adhere to the visual grid. 
- It adds a new popover prop, "isPrimary", which puts the popover visually at elevation zero and shows dark borders. This is meant to be used with the block toolbar, or _when you know what you're doing_, and by default the popover still shows a gray border and drop shadow.

Aside from the visual considerations, I'd like feedback on the popover prop code, I'm sure I could've done that in a nicer way. 

I would also love some RTL feedback on the default popover sirection, to see if we can maybe change the default per locale.

jasmussen committed:
Try removing the triangle indicator.

jasmussen committed:
Remove the gray fade again.

jasmussen committed:
Fix busy gradient.

jasmussen commented:
A few screenshots of work in progress. 

I tried removing the triangle, because we might not be able do the negative leftwards shift of the toolbar. Before:

<img width="810" alt="triangle-before" src="https://user-images.githubusercontent.com/1204802/78760023-d2585300-7980-11ea-9fb7-dff36cca1c37.png">

After:

<img width="725" alt="triangle-afterafter" src="https://user-images.githubusercontent.com/1204802/78760033-d6847080-7980-11ea-8136-71ab0a39fc76.png">

I fixed the saving gradient. Before:

![before gradient](https://user-images.githubusercontent.com/1204802/78760052-ddab7e80-7980-11ea-8935-c33813ad41e7.gif)

After:

![after gradient](https://user-images.githubusercontent.com/1204802/78760061-e00dd880-7980-11ea-8cfc-65c029835076.gif)

jasmussen committed:
Update latest post icon.

jasmussen committed:
Tweak active state for alignments.

jasmussen committed:
Also do that for block navigation.

jasmussen commented:
Tweaked alignments. Before:

<img width="397" alt="Screenshot 2020-04-08 at 10 18 49" src="https://user-images.githubusercontent.com/1204802/78762361-0ed97e00-7984-11ea-98d4-7c459085a931.png">

After:

<img width="438" alt="Screenshot 2020-04-08 at 10 29 28" src="https://user-images.githubusercontent.com/1204802/78762450-27e22f00-7984-11ea-9c92-5f6213f865a5.png">

Also did this for block navigation:

<img width="303" alt="Screenshot 2020-04-08 at 10 34 26" src="https://user-images.githubusercontent.com/1204802/78762762-94f5c480-7984-11ea-934e-493f6e90efee.png">

jasmussen committed:
Animate focus state.

jasmussen committed:
Unfloat notice button.

Fixes #21316.

jasmussen commented:
Added animation to toggle buttons:

![bold appear](https://user-images.githubusercontent.com/1204802/78764295-a344e000-7986-11ea-8770-a4c13c488d2e.gif)

Unfloated the notice button to fix an issue it regressed. These notices should be redesigned separately:

<img width="1351" alt="Screenshot 2020-04-08 at 10 48 52" src="https://user-images.githubusercontent.com/1204802/78764348-b5bf1980-7986-11ea-8e37-5a200e6a3f89.png">

jasmussen committed:
Popovers: default to bottom right.

Fixes #20765.

jasmussen commented:
Pushed a fix to set a default popover direction. This adds some much needed consistency:

![down right](https://user-images.githubusercontent.com/1204802/78766348-71814880-7989-11ea-907d-3054720d5e53.gif)

jasmussen committed:
Add spacing to popover.

jasmussen committed:
Add a "primary" prop to popover, and apply to menu popovers.

jasmussen commented:
I added spacing to popovers, and a new "isPrimary" prop to the popover which will paint it dark and at elevation zero:

![primary submenus](https://user-images.githubusercontent.com/1204802/78772892-b8c00700-7992-11ea-9d6d-5b9d92afb189.gif)

This gets us very close to the mockups. 

By default, the popover still uses the light gray shadowed treatment:

<img width="311" alt="Screenshot 2020-04-08 at 12 16 35" src="https://user-images.githubusercontent.com/1204802/78772958-d0978b00-7992-11ea-9a85-98b7a35547e9.png">

The primary prop should only be used when the popover is in the context of block toolbar UI.

ZebulanStanphill commented:
I like the new popover styles.

>The primary prop should only be used when the popover is in the context of block toolbar UI.

Why?

jasmussen commented:
I suppose I should've finished that sentence with "... or if you know what you're doing".

I think of it as an extension of the block toolbar, being made of the same material and being coplanar with it: extending at the same z depth layer and therefore subject to the same "physics".

When a popover is opened from the top toolbar, on the other hand, it's not the same material and it will actually overlay the toolbar itself and often, the sidebar. That beckons a higher z depth, which suggests the shadow. Extending from secondary (to the block) UI also means the dark border wouldn't connect in any meaningful way.

That's how I think about it, in any case!

ellatrix commented:
Love the improvements!

What if the popover outlines were the same, and the popover was positioned slightly lower and separate from the toolbar? Probably discussed before, so sorry if I missed that. :)


![](https://user-images.githubusercontent.com/1204802/78762450-27e22f00-7984-11ea-9c92-5f6213f865a5.png)

jasmussen commented:
@ellatrix Thanks!

> What if the popover outlines were the same, and the popover was positioned slightly lower and separate from the toolbar? Probably discussed before, so sorry if I missed that. :)

But that is exactly what this PR does, or am I misunderstanding?

<img src="https://user-images.githubusercontent.com/1204802/78772892-b8c00700-7992-11ea-9d6d-5b9d92afb189.gif">

ellatrix commented:
Ooooh, perfect then!

ellatrix commented:
> * It adds a new popover prop, "isPrimary", which puts the popover visually at elevation zero and shows dark borders. This is meant to be used with the block toolbar, or _when you know what you're doing_, and by default the popover still shows a gray border and drop shadow.

Ah, I missed this bit. Is there a need to differentiate? Why not make all popovers like this?

jasmussen commented:
>  Is there a need to differentiate? Why not make all popovers like this?

I think so, yes. I went into a little more depth in my comment to Zeb (https://github.com/WordPress/gutenberg/pull/21476#issuecomment-611026657) — but it boils down to this: by default the popover component should behave as a popover: it _overlays content_.

But in the specific case of popovers that extend from the block toolbar, these are so literally close to the block itself, that they are at what I would call "elevation zero". Not talking CSS z-index here, but rather something similar to Material's "elevation" principle: https://material.io/design/environment/elevation.html

Another way to think of it is just that those popovers just receive more emphasis because they are primary UI. Same principle as when you have a single Button component with an isPrimary attribute.

mtias commented on packages/block-editor/src/components/alignment-toolbar/index.js:
```diff
@@ -30,6 +30,7 @@ const DEFAULT_ALIGNMENT_CONTROLS = [
 
 const POPOVER_PROPS = {
 	position: 'bottom right',
+	isPrimary: true,
```
Wonder if we should use a more descriptive name here because I don't feel primary / secondary is what should determine the weight. Maybe something like "isFloating"? Or "isBlockUi" since these are used for blocks only? Just a thought.

jasmussen replied:
I'm super duper open to changing this, and now is the best time to do so. Which is your preference? We could also go in a different direction, like `isEmphasized`?

mtias replied:
Maybe `isAlternate` for now?

ZebulanStanphill commented:
It seems weird to me to use shadows in some popovers but not others. When I first saw the G2 mockups, I thought we were moving away from shadows entirely and that everything would use the high contrast borders that the block toolbar uses.

It seems weird to me to use shadows in some places but not others. If some popovers are going to have shadows, why not also the toolbar? I would think of it as being at a higher elevation than the content being edited, yet it has no shadows. It seems like an inconsistency to me to use a mix of some floating elements having shadows and others not.

Also, I don't really get the idea of emphasizing certain popovers over others. All the popovers are closed by default, so different styling on one versus the other isn't really going to make a difference on where your eyes are drawn to. When you open the more menu in the top toolbar, your attention is focused on the more menu popover, so shouldn't it appear just as emphasized as any other popover?

I'm not convinced other users are going to understand why some popovers look different than others, either. Without having seen your comments, I would have assumed it was just a visual consistency bug.

That said, if others disagree with me, I'm okay with this PR merging as-is, design-wise. I'm happy to have the popovers looking more consistent with the block toolbar UI, even if it's only some of the popovers.

jasmussen commented:
> It seems weird to me to use shadows in some popovers but not others. When I first saw the G2 mockups, I thought we were moving away from shadows entirely and that everything would use the high contrast borders that the block toolbar uses.

To quote a friend of mine: consistency is nuanced.

Honestly, the term "popover" is not one a user is likely ever going to see or even think about, it's just what the underlying component is called. The top toolbar and UI outside the canvas, the borders and separators are intentionally lighter because it is secondary UI and because it does not have to visually distinguish itself from the theme. Menus that extend from this UI are consistent with this UI. 

And inside the canvas, there's the block toolbar. This UI has been designed not only to be primary, but to work in any context background:

![dark-theme](https://user-images.githubusercontent.com/1204802/78965855-8c28fe00-7afe-11ea-8f40-8136056a3142.png)

![text-on-backgrounds](https://user-images.githubusercontent.com/1204802/78965858-8d5a2b00-7afe-11ea-98cd-843a49517c72.png)

Yes, technically alignments appear in a popover — but just like the "Parent Button" is made from toolbar material, so are the menus that extend from the toolbar.

jasmussen committed:
isAlternate instead of isPrimary

jasmussen committed:
Smaller plus for appender.

jasmussen commented:
Renamed isPrimary prop to isAlternate. Also used the smaller plus for the big appender button. Looks good, props Pablo for tip:

<img width="898" alt="Screenshot 2020-04-10 at 17 11 25" src="https://user-images.githubusercontent.com/1204802/79001367-a25eaa80-7b4e-11ea-872d-2c36bb749277.png">

jasmussen committed:
Update snapshots.

jasmussen commented:
Addressed feedback, rebased, updated snapshots. 

I feel like this branch is in a solid solid place, and none of the disagreements feel like ones we can't revisit later on, if they turn out to be problematic. I would definitely still welcome thoughts and such, and most definitely testing. But otherwise, this is ready for final review!

ZebulanStanphill reviewed (approved):
Looks better than `master` to me. :+1:

jasmussen commented:
Thank you Zeb. I forgot to test again RTL and the new default popover direction, so I'll test and verify that before merging.

jasmussen commented:
I wanted to make sure that the new default behavior for popovers to open down and rightwards was properly handled in RTL language directions. It is, it defaults to down and leftwards:

![menus](https://user-images.githubusercontent.com/1204802/79097118-31421180-7d5f-11ea-9dad-46bce748a04d.gif)

In that vein, I'm going to merge this one. 

I wanted to make a note that the change in https://github.com/WordPress/gutenberg/pull/21476#issuecomment-610816232 to remove the triangle indicator is one that is worth revisiting if/when we look at the mover control behavior again. I consider it a delightful aspect of the G2 visuals, and one that can provide great character when it does indeed point to the left margin of the block. So I would love to restore it when we can.

mtias commented:
Nice work, Joen!
